src/components/tl/mlx5/alltoall/alltoall_coll.c

```diff
@@ -0,0 +1,873 @@
+        if (status != UCC_OK) {
+            tl_error(UCC_TASK_LIB(task), "Failed sending atomic to node [%d]",
+                     cyc_rank);
+            return status;
+        }
+    }
+    if (!task->alltoall.send_blocks_enqueued) {
+        ucc_progress_enqueue(UCC_TL_CORE_CTX(team)->pq, coll_task);
+        task->alltoall.send_blocks_enqueued = 1;
+    }
+
+    if (task->alltoall.started == a2a->net.net_size) {
+        status = ucc_tl_mlx5_post_wait_on_data(
+            a2a->net.umr_qp, a2a->net.net_size, a2a->net.atomic.mr->lkey,
+            (uintptr_t)
+#if ATOMIC_IN_MEMIC
+                PTR_OFFSET(0,
+#else
+                PTR_OFFSET(a2a->net.atomic.counters,
+#endif
+                           seq_index * sizeof(tl_mlx5_atomic_t)),
+            task);
+    }
+
+    return status;
+}
+
+static void ucc_tl_mlx5_send_blocks_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_mlx5_schedule_t *task = TASK_SCHEDULE(coll_task);
+    ucc_tl_mlx5_team_t     *team = SCHEDULE_TEAM(task);
+    ucc_tl_mlx5_alltoall_t *a2a  = team->a2a;
+    ucc_status_t            status;
+
+    if (task->alltoall.started != a2a->net.net_size) {
+        coll_task->post(coll_task);
+        return;
+    }
+    status = ucc_tl_mlx5_poll_cq(a2a->net.cq, UCC_TASK_LIB(coll_task));
+    if (UCC_OK != status) {
+        coll_task->status = status;
+        return;
+    }
+
+    if (task->alltoall.blocks_sent == task->alltoall.blocks_completed) {
+        UCC_TL_MLX5_PROFILE_REQUEST_EVENT(
+            task, "mlx5_alltoall_all_blocks_completed", 0);
+        coll_task->status = UCC_OK;
+    }
+}
+
+ucc_status_t ucc_tl_mlx5_alltoall_start(ucc_coll_task_t *coll_task)
+{
+    UCC_TL_MLX5_PROFILE_REQUEST_EVENT(coll_task, "mlx5_alltoall_start", 0);
+    return ucc_schedule_start(coll_task);
+}
+
+ucc_status_t ucc_tl_mlx5_alltoall_finalize(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_mlx5_schedule_t *task =
+        ucc_derived_of(coll_task, ucc_tl_mlx5_schedule_t);
+    ucc_tl_mlx5_team_t     *team      = SCHEDULE_TEAM(task);
+    ucc_tl_mlx5_alltoall_t *a2a       = team->a2a;
+    ucc_tl_mlx5_context_t  *ctx       = SCHEDULE_CTX(task);
+    int                     seq_index = task->alltoall.seq_index;
+    ucc_status_t            status;
+
+    a2a->previous_msg_size[seq_index] = task->alltoall.msg_size;
+    a2a->previous_send_address[seq_index] =
+        task->alltoall.send_rcache_region_p->reg.mr->addr;
+    a2a->previous_recv_address[seq_index] =
+        task->alltoall.recv_rcache_region_p->reg.mr->addr;
+    ucc_rcache_region_put(ctx->rcache,
+                          &task->alltoall.send_rcache_region_p->super);
+    ucc_rcache_region_put(ctx->rcache,
+                          &task->alltoall.recv_rcache_region_p->super);
+    UCC_TL_MLX5_PROFILE_REQUEST_EVENT(coll_task, "mlx5_alltoall_done", 0);
+    status = ucc_schedule_finalize(coll_task);
+    ucc_tl_mlx5_put_schedule(task);
+    return status;
+}
+
+static inline int power2(int value)
```

Comment:
No, I don't think it can be expressed in terms of `ucc_align_down_pow2`. We actually already evoked this [here](https://github.com/openucx/ucc/pull/724#discussion_r1127370329) 
In my opinion this is the only/best way to do, with a while loop like this